### PR TITLE
MINOR: Reduce memory allocation in ClientTelemetryReporter.java

### PR DESCRIPTION
**Pull request by testn**
ClientTelemetryReporter.timeToNextUpdate is called really often when sending a lot of data to Kafka servers. This is due to its logging. It tries to log the current state by formatting a string which generates a lot of char[] underneath.

This change will make it such that it will not generate string if the tracing is disabled.

### Committer Checklist (excluded from commit message)
- [X] Verify design and implementation 
- [X] Verify test coverage and CI build status
- [X] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by divijvaidya — APPROVED** [review]
Nice! Small but very helpful change.

Out of curiosity, do you happen to have any ideas to detect such performance-inefficiencies automatically in future?

**Comment by divijvaidya** [comment]
@testn could you please fix the checkstyle failures which is causing the build to fail for this PR. https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-15402/2/pipeline

**Comment by testn** [comment]
Add the file to the suppression list instead.

**Comment by testn** [comment]
> Nice! Small but very helpful change.
> 
> Out of curiosity, do you happen to have any ideas to detect such performance-inefficiencies automatically in future?

Currently, kafka is not really optimized yet. There are too many areas of improvements we can achieve especially from the memory usage perspective.

File: checkstyle/suppressions.xml [hunk 1]
```diff
@@ -92,7 +92,7 @@
               files="(Utils|Topic|KafkaLZ4BlockOutputStream|AclData|JoinGroupRequest).java"/>
 
     <suppress checks="CyclomaticComplexity"
-              files="(AbstractFetch|ConsumerCoordinator|CommitRequestManager|FetchCollector|OffsetFetcherUtils|KafkaProducer|Sender|ConfigDef|KerberosLogin|AbstractRequest|AbstractResponse|Selector|SslFactory|SslTransportLayer|SaslClientAuthenticator|SaslClientCallbackHandler|SaslServerAuthenticator|AbstractCoordinator|TransactionManager|AbstractStickyAssignor|DefaultSslEngineFactory|Authorizer|RecordAccumulator|MemoryRecords|FetchSessionHandler|MockAdminClient).java"/>
+              files="(AbstractFetch|ClientTelemetryReporter|ConsumerCoordinator|CommitRequestManager|FetchCollector|OffsetFetcherUtils|KafkaProducer|Sender|ConfigDef|KerberosLogin|AbstractRequest|AbstractResponse|Selector|SslFactory|SslTransportLayer|SaslClientAuthenticator|SaslClientCallbackHandler|SaslServerAuthenticator|AbstractCoordinator|TransactionManager|AbstractStickyAssignor|DefaultSslEngineFactory|Authorizer|RecordAccumulator|MemoryRecords|FetchSessionHandler|MockAdminClient).java"/>
```
Comment (divijvaidya) [inline checkstyle/suppressions.xml:96]: why are we suppressing this? Ideally we would like to follow the same styling as rest of the code and hence fix the checkstyle failures.

Comment (testn) [reply checkstyle/suppressions.xml:96]: How about removing the trace out completely? I'm not a fan of calling trace in such a tight loop

Comment (divijvaidya) [reply checkstyle/suppressions.xml:96]: I am not sure who might be relying on this log and I am not aware of the original motivation of adding this. I would prefer to keep the current state and only focus on perf improvement in this PR.

**Review by AndrewJSchofield — APPROVED** [review]

File: checkstyle/suppressions.xml (hunk 1 above)
Comment (testn) [reply checkstyle/suppressions.xml:96]: If you prefer to keep the trace logic, the added code is purely to reduce memory allocation but it comes with extra baggage for additional cyclomatic complexity and hence the suppression

Comment (divijvaidya) [reply checkstyle/suppressions.xml:96]: Understood. Thanks. This answers my question on my we are suppressing it.

**Comment by divijvaidya** [comment]
The build is still failing though @testn please check.

**Comment by testn** [comment]
@divijvaidya  seem like the build is orange!

image

**Comment by divijvaidya** [comment]
The test failures are unrelated to this change.

@testn You can see the test failures in the orange build by clicking on the arrow on top right. We have a flaky test suite but if a test is failing consistently, it usually is caused by the change in the PR. In this case, things look ok.

**Comment by divijvaidya** [comment]
@testn congratulations on your first contribution to Apache Kafka! Hopefully first of many more to come.

File: clients/src/main/java/org/apache/kafka/common/telemetry/internals/ClientTelemetryReporter.java [hunk 2]
```diff
@@ -336,33 +337,35 @@ public long timeToNextUpdate(long requestTimeoutMs) {
...
                         timeMs = timeRemainingBeforeRequest;
-                        msg = String.format("the client will wait before submitting the next %s network API request", apiName);
+                        msg = isTraceEnabled ? "" : String.format("the client will wait before submitting the next %s network API request", apiName);
                     }
                     break;
                 default:
                     throw new IllegalStateException("Unknown telemetry state: " + localState);
             }
 
-            log.trace("For telemetry state {}, returning the value {} ms; {}", localState, timeMs, msg);
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/telemetry/internals/ClientTelemetryReporter.java:365]: Hmm, unlike what the PR says, there is no string generated if the log is disabled. The only allocation is the array for passing the parameters. No harm in this change, but I wanted to clarify.

Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/telemetry/internals/ClientTelemetryReporter.java:365]: More details here https://www.baeldung.com/java-string-pool

**Pull request merged**